Inject X-Smartstack-Source header in Casper

yannlandry opened the pull request:
This incorporates @bplotnick's work which I haven't touched yet. Maybe we need to not-hardcode the `spectre.main` value that is being injected? I see that HAProxy gets a hold of a `proxied_through` value to inject the same header but when we're in Casper we know it's gonna be `spectre.main`.

Test-wise, I've gotten around to figure out how Casper does tests and itests and it turns out we use a fake backend behind Casper to return whatever was sent to it, so I just make a request that doesn't have a `X-Smartstack-Source` header and check if the backend received one.

ArnaudBrousseau commented on lua/spectre_common.lua:
```diff
@@ -337,9 +336,6 @@ end
 -- Returns an error message as a string if a malformed request is detected.
 -- --
 -- To be proxied through spectre, a request needs 2 headers
```
```suggestion
-- To be proxied through Casper, a request needs the following headers:
```

ArnaudBrousseau commented on lua/spectre_common.lua:
```diff
@@ -337,9 +336,6 @@ end
---   request. This header is set and used by HAProxy to proxy requests through
---   Spectre exactly once
 -- + X-SmartStack-Destination: the nerve namespace of the service called. This
 --   lets Spectre lookup the relevant set of configs and forward to the right
```
```suggestion
--   lets Casper lookup the relevant set of configs and forward to the right
```

ArnaudBrousseau commented on tests/lua/spectre_common_test.lua:
```diff
@@ -361,19 +360,17 @@ describe("spectre_common", function()
                 assert.are.equal(false, should_proxy)
                 assert.are.equal('X-Smartstack-Destination has multiple values: dst1 dst2;', err)
             end)
             it("Combines error messages when multiple sources AND destinations are provided", function()
```
I'd just remove this test since it now does the exact same thing than the one above it.

drolando commented on lua/spectre_common.lua:
```diff
@@ -319,6 +314,10 @@ local function forward_to_destination(method, request_uri, request_headers)
 end
 
+local function inject_source_header(request)
+   request.set_header('X-Smartstack-Source', 'spectre.main')
```
I don't think you can use `PAASTA_SERVICE` or `PAASTA_SERVICE`. The instances are called main or canary in all environments. I'd also rather have it in the config file rather than assuming this runs of paasta.

What I don't remember is if this is `spectre.main` also in dev and stage or if there we use `spectre.dev` and `spectre.stage`

drolando commented on lua/spectre_common.lua:
```diff
@@ -319,6 +314,10 @@ local function forward_to_destination(method, request_uri, request_headers)
     }
 end
 
+local function inject_source_header(request)
```
Missing docstring. Please mention that we're doing this so that haproxy/envoy knows that this request is from casper and won't re-route it back to it

yannlandry commented:
Okay, so I've fixed all PR comments.

Casper reads the value from a config file for `X-Smartstack-Source` and I've also added a unit test for that.

`make test` and `make itest` pass.

drolando commented on lua/spectre_common.lua:
```diff
@@ -314,8 +314,12 @@ local function forward_to_destination(method, request_uri, request_headers)
 local function inject_source_header(request)
-   request.set_header('X-Smartstack-Source', 'spectre.main')
+    local configs = config_loader.get_spectre_config_for_namespace(config_loader.CASPER_INTERNAL_NAMESPACE)
+    local x_smartstack_source_value = configs['casper']['x_smartstack_source_value']
```
+1

yannlandry commented:
All right so I've added a default value + a unit test to cover that. The unit test is a little bit ugly since I had to manually replace then restore the `config_loader.get_spectre_config_for_namespace` function but according to my research it's the only way to currently do it with that test framework.
